lib/CodeGen/PagerandoBinning.cpp

```diff
@@ -72,45 +72,56 @@ ModulePass *llvm::createPagerandoBinningPass() {
   return new PagerandoBinning();
 }
 
-static unsigned GetFunctionSizeInBytes(const MachineFunction &MF, const TargetInstrInfo *TII) {
-  unsigned FnSize = 0;
-  for (auto &MBB : MF)
-    for (auto &MI : MBB)
-      FnSize += TII->getInstSizeInBytes(MI);
-  return FnSize;
-}
-
 bool PagerandoBinning::runOnModule(Module &M) {
   MachineModuleInfo &MMI = getAnalysis<MachineModuleInfo>();
 
-  // Bin all functions
   for (auto &F : M) {
-    if (!F.isRandPage()) {
-      // Put all normal functions (and wrappers) into bin 0.
-      MMI.setBin(&F, 0);
-      continue;
-    }
-
-    MachineFunction &MF = MMI.getMachineFunction(F);
-    const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
-
-    unsigned FnSize = GetFunctionSizeInBytes(MF, TII);
+    unsigned Bin = F.isRandPage() ? AssignToBin(F) : DefaultBin;
+    MMI.setBin(&F, Bin);
+  }
 
-    auto I = Bins.lower_bound(FnSize);
-    if (I == Bins.end())
-      I = Bins.emplace(Bin::BinSize, BinCount++);
+  return true;
+}
 
-    // Add the function to the given bin
-    DEBUG(dbgs() << "Putting function '" << MF.getName() << "' with size " << FnSize << " in bin " << I->second.Number << " with free space " << I->second.FreeSpace << '\n');
-    MMI.setBin(&F, I->second.Number);
-    if (FnSize <= I->second.FreeSpace)
-      I->second.FreeSpace -= FnSize;
-    else
-      I->second.FreeSpace = (FnSize - I->second.FreeSpace) % Bin::BinSize;
-    Bins.insert(std::make_pair(I->second.FreeSpace, I->second));
+unsigned PagerandoBinning::AssignToBin(const Function &F) {
+  unsigned FnSize = ComputeFunctionSize(F);
+  unsigned Bin, FreeSpace;
+
+  auto I = Bins.lower_bound(FnSize);
+  if (I == Bins.end()) {  // No bin with enough free space
+    Bin = BinCount++;
+    if (FnSize % BinSize == 0) { // Function size is a multiple of bin size
+      FreeSpace = 0;
+    } else {
+      FreeSpace = BinSize - (FnSize % BinSize);
+    }
+  } else {                // Found eligible bin
+    Bin = I->second;
+    FreeSpace = I->first - FnSize;
     Bins.erase(I);
   }
 
-  return true;
+  if (FreeSpace >= MinFreeSpace) {
+    Bins.emplace(FreeSpace, Bin);
+  }
+
+  DEBUG(dbgs() << "Assigning function '" << F.getName()
+               << "' with size " << FnSize
+               << " to bin #" << Bin
+               << " with remaining free space " << FreeSpace << '\n');
+
+  return Bin;
 }
 
+unsigned PagerandoBinning::ComputeFunctionSize(const Function &F) {
+  const MachineFunction &MF = getAnalysis<MachineModuleInfo>().getMachineFunction(F);
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+
+  unsigned Size = 0;
+  for (auto &MBB : MF)
+    for (auto &MI : MBB)
+      Size += TII->getInstSizeInBytes(MI);
+
+  assert(Size > 0 && "Function size is assumed to be greater than zero.");
```

Comment:
If this doesn't hold (function size cannot be determined), we could assume a default size for functions? Otherwise I have to think again how this affects binning.